hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -20,62 +20,159 @@
+        The lower limit for max_element_length is 1 in accordance with :rfc:`1035#section-2.3.4`
+        """
+        # https://tools.ietf.org/html/rfc1035#section-2.3.4
+
+        max_length = self.clean_inputs(4, 255, max_length, "max_length")
+        max_element_length = self.clean_inputs(
+            1, 63, max_element_length, "max_element_length"
+        )
+
+        SearchStrategy.__init__(self)
+        self.max_length = max_length
+        self.max_element_length = max_element_length
+
+        if self.max_element_length == 1:
+            self.label_regex = r"[a-zA-Z]"
+        elif self.max_element_length == 2:
+            self.label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
+        else:
+            self.label_regex = (
+                r"[a-zA-Z]([a-zA-Z0-9\-]{0,"
+                + str(self.max_element_length - 2)
+                + r"}[a-zA-Z0-9])?"
+            )
+
+    def do_draw(self, data):
+        # Select a valid top-level domain (TLD) name,
+        # then randomize it between upper and lower case characters.
+        # Finally, check that our TLD
+        domain = data.draw(
+            st.sampled_from(TOP_LEVEL_DOMAINS)
+            .flatmap(
+                lambda tld: st.tuples(
+                    *[st.sampled_from([c.lower(), c.upper()]) for c in tld]
+                ).map(u"".join)
+            )
+            .filter(lambda tld: len(tld) + 2 <= self.max_length)
+        )
+        # The maximum possible subdomains is 126,
+        # 1 letter sub-domains + 1 '.' character, * 126 = 252,
+        # with a max of 255, that leaves 3 characters for a TLD.
+        elements = cu.many(data, min_size=1, average_size=1, max_size=126)
+        while elements.more():
+            # Generate a new valid sub-domain using the regex strategy.
+            sub_domain = data.draw(st.from_regex(self.label_regex, fullmatch=True))
+            if len(domain) + len(sub_domain) >= self.max_length:
+                data.stop_example(discard=True)
+                break
+            domain = sub_domain + "." + domain
+        return domain
+
+
+@st.defines_strategy_with_reusable_values
+def domains(
+    max_length=255,  # type: int
+    max_element_length=63,  # type: int
+):
+    # type: (...) -> SearchStrategy[Text]
+    """Generate :rfc:`1035` compliant fully qualified domain names."""
+    return DomainNameStrategy(
+        max_length=max_length, max_element_length=max_element_length
+    )
 
 
 @st.defines_strategy_with_reusable_values
 def urls():
+    # type: () -> SearchStrategy[Text]
     """A strategy for :rfc:`3986`, generating http/https URLs."""
 
     def url_encode(s):
-        safe_chars = set(string.ascii_letters + string.digits + "$-_.+!*'(),")
-        return "".join(c if c in safe_chars else "%%%02X" % ord(c) for c in s)
+        return "".join(c if c in URL_SAFE_CHARACTERS else "%%%02X" % ord(c) for c in s)
 
-    schemes = st.sampled_from(["http", "https"])
+    schemes = st.sampled_from(["", "s"])
```

Comment:
nit: trying to follow your new code, I prefer the old way – it’s really not clear what the empty string/`"s"` mean in a way that `"http"`/`"https"` is.

Reply:
This was one of the changes @Zac-HD included in his work, so I’ll defer to him for an explanation why to keep it, otherwise I’m happy to swap back cause I agree with you @alexwlchan

Reply:
Swap it back; that was a style change when I was fiddling with it but I prefer to go with the majority view :smile:

Reply:
fixed in my latest changes. :+1: